### PR TITLE
Issue #11420 - fix dynamic table referencing in QpackDecoder

### LINKED ISSUE #11420: HTTP/3 Race with Dynamic Table referencing in QPackDecoder
**Jetty version(s)**
12.0.7

**Jetty Environment**
core

**Java version/vendor** `(use: java -version)`
openjdk version "21.0.1" 2023-10-17 LTS
OpenJDK Runtime Environment Zulu21.30+15-CA (build 21.0.1+12-LTS)
OpenJDK 64-Bit Server VM Zulu21.30+15-CA (build 21.0.1+12-LTS, mixed mode, sharing)

**OS type/version**
Microsoft Windows [Version 10.0.19045.4046]

**Description**
I have an `index.html` file that references `login/login.js`.

If I open a new Incognito tab and hit `index.html` I sometimes (randomly) get the wrong headers.

Here is what a good request looks like:

```
14:37:32.330 [main-21] DEBUG o.e.j.h.s.i.HttpStreamOverHTTP3.onRequest() - HTTP3 request #4/2327826b, GET https://licensed.app/login/login.js HTTP/3.0
sec-ch-ua: "Not A(Brand";v="99", "Brave";v="121", "Chromium";v="121"
origin: https://licensed.app
dnt: 1
sec-ch-ua-mobile: ?0
user-agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/121.0.0.0 Safari/537.36
sec-ch-ua-platform: "Windows"
accept: */*
sec-gpc: 1
sec-fetch-site: same-origin
sec-fetch-mode: cors
sec-fetch-dest: script
referer: https://licensed.app/
accept-encoding: gzip, deflate, br
accept-language: en-US,en;q=0.9
```

And here is what a bad request looks like:

```
14:37:47.395 [main-21] DEBUG o.e.j.h.s.i.HttpStreamOverHTTP3.onRequest() - HTTP3 request #56/2f56978, GET https://licensed.app/login/login.js HTTP/3.0
pragma: no-cache
cache-control: no-cache
sec-ch-ua: "Not A(Brand";v="99", "Brave";v="121", "Chromium";v="121"
origin: https://licensed.app
dnt: 1
sec-ch-ua-mobile: ?0
user-agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/121.0.0.0 Safari/537.36
sec-ch-ua-platform: "Windows"
accept: */*
sec-gpc: 1
sec-gpc: same-origin
sec-gpc: cors
accept: script
referer: https://licensed.app/
accept-encoding: gzip, deflate, br
accept-language: en-US,en;q=0.9
```

If you diff these two, you will notice that:

```
sec-fetch-site: same-origin
sec-fetch-mode: cors
sec-fetch-dest: script
```

is replaced by:

```
sec-gpc: same-origin
sec-gpc: cors
accept: script
```

The bad request also contains:

```
pragma: no-cache
cache-control: no-cache
```

which the good request does not. The only reason I caught this is that `accept: script` is invalid.

It's not clear to me whether this is caused by a race condition in Jetty (possibly only in HTTP/3 mode) or it's caused by a race race condition on Chrome.

I know that the initial (non-cached) page load is always okay. If I get the bad headers, it is typically on the second (and subsequent) requests when I hit F5 to reload the page.

Because this is a race condition, the more logs I add the less likely I am to run into this problem. I cannot enable all the DEBUG logs.

How do you suggest figuring out which side is at fault?

### PR DESCRIPTION
**Pull request by lachlan-roberts**
## Issue #11420

Some places in the QPACK decoding were using absolute index instead of relative index.

### CODE REVIEW

File: jetty-core/jetty-http3/jetty-http3-qpack/src/main/java/org/eclipse/jetty/http3/qpack/internal/QpackContext.java [hunk 1]
```diff
@@ -66,6 +66,7 @@ public Entry get(String name)
         return _dynamicTable.get(StringUtil.asciiToLowerCase(name));
     }
 
+    @Deprecated
```
Comment (sbordet) [inline jetty-core/jetty-http3/jetty-http3-qpack/src/main/java/org/eclipse/jetty/http3/qpack/internal/QpackContext.java]: Remove it, it's in an internal package.

File: jetty-core/jetty-http3/jetty-http3-qpack/src/test/java/org/eclipse/jetty/http3/qpack/EvictionTest.java [hunk 2]
```diff
@@ -33,7 +33,7 @@ public class EvictionTest
     private QpackDecoder _decoder;
     private final TestDecoderHandler _decoderHandler = new TestDecoderHandler();
     private final TestEncoderHandler _encoderHandler = new TestEncoderHandler();
-    private final Random random = new Random();
+    private final Random random = new Random(10);
```
Comment (sbordet) [inline jetty-core/jetty-http3/jetty-http3-qpack/src/test/java/org/eclipse/jetty/http3/qpack/EvictionTest.java]: Get rid of seed.

File: jetty-core/jetty-http3/jetty-http3-qpack/src/main/java/org/eclipse/jetty/http3/qpack/internal/table/DynamicTable.java [hunk 3]
```diff
@@ -164,9 +156,24 @@ public Entry get(HttpField field)
...
+    {
+        return getRelative(index, getInsertCount());
+    }
+
+    public Entry getRelative(int index, int base)
+    {
+        return getAbsolute(base - 1 - index);
+    }
+
+    public Entry getPostBase(int index, int base)
```
Comment (sbordet) [inline jetty-core/jetty-http3/jetty-http3-qpack/src/main/java/org/eclipse/jetty/http3/qpack/internal/table/DynamicTable.java]: Can you switch the parameters order to be `base, index`? Feels more natural.

File: jetty-core/jetty-http3/jetty-http3-qpack/src/main/java/org/eclipse/jetty/http3/qpack/internal/table/DynamicTable.java [hunk 4]
```diff
@@ -164,9 +156,24 @@ public Entry get(HttpField field)
         return _fieldMap.get(field);
     }
 
+    public Entry getRelative(int index)
+    {
+        return getRelative(index, getInsertCount());
+    }
+
+    public Entry getRelative(int index, int base)
```
Comment (sbordet) [inline jetty-core/jetty-http3/jetty-http3-qpack/src/main/java/org/eclipse/jetty/http3/qpack/internal/table/DynamicTable.java]: Ditto.

**Review by sbordet — CHANGES_REQUESTED** [review]
Can you please add a full client/server test that reproduces the issue?
Just to be sure that the interactions of the various components in H3 works as expected.

File: jetty-core/jetty-http3/jetty-http3-qpack/src/main/java/org/eclipse/jetty/http3/qpack/internal/QpackContext.java (hunk 1 above)
Comment (lachlan-roberts) [reply jetty-core/jetty-http3/jetty-http3-qpack/src/main/java/org/eclipse/jetty/http3/qpack/internal/QpackContext.java]: done

File: jetty-core/jetty-http3/jetty-http3-qpack/src/test/java/org/eclipse/jetty/http3/qpack/EvictionTest.java (hunk 2 above)
Comment (lachlan-roberts) [reply jetty-core/jetty-http3/jetty-http3-qpack/src/test/java/org/eclipse/jetty/http3/qpack/EvictionTest.java]: done

File: jetty-core/jetty-http3/jetty-http3-qpack/src/main/java/org/eclipse/jetty/http3/qpack/internal/table/DynamicTable.java (hunk 3 above)
Comment (lachlan-roberts) [reply jetty-core/jetty-http3/jetty-http3-qpack/src/main/java/org/eclipse/jetty/http3/qpack/internal/table/DynamicTable.java]: done

File: jetty-core/jetty-http3/jetty-http3-qpack/src/main/java/org/eclipse/jetty/http3/qpack/internal/table/DynamicTable.java (hunk 4 above)
Comment (lachlan-roberts) [reply jetty-core/jetty-http3/jetty-http3-qpack/src/main/java/org/eclipse/jetty/http3/qpack/internal/table/DynamicTable.java]: done

**Review by sbordet — APPROVED** [review]

**Pull request merged**